libs/mbedtls/mbedtls_stubs.c

```diff
@@ -0,0 +1,307 @@
+#include <ctype.h>
+#include <string.h>
+#include <stdio.h>
+
+#ifdef _WIN32
+#include <windows.h>
+#include <wincrypt.h>
+#endif
+
+#include <caml/mlvalues.h>
+#include <caml/alloc.h>
+#include <caml/memory.h>
+#include <caml/fail.h>
+#include <caml/callback.h>
+
+#include "mbedtls/debug.h"
+#include "mbedtls/error.h"
+#include "mbedtls/config.h"
+#include "mbedtls/ssl.h"
+#include "mbedtls/entropy.h"
+#include "mbedtls/ctr_drbg.h"
+#include "mbedtls/net_sockets.h"
+#include "mbedtls/certs.h"
+
+#define SSL(v) (mbedtls_ssl_context*)Field(v, 0)
+#define CFG(v) (mbedtls_ssl_config*)Field(v, 0)
+#define CTR(v) (mbedtls_ctr_drbg_context*)Field(v, 0)
+#define CERT(v) (mbedtls_x509_crt*)Field(v, 0)
+#define PVOID(v) (void*)Field(v, 0)
+
+void debug(void* ctx, int debug_level, const char* file_name, int line, const char* message) {
+	printf("%s:%i: %s", file_name, line, message);
+}
+
+CAMLprim value ml_mbedtls_strerror(value code) {
+	CAMLparam1(code);
+	CAMLlocal1(r);
+	char buf[128];
+	mbedtls_strerror(Int_val(code), buf, sizeof(buf));
+	r = caml_copy_string(buf);
+	CAMLreturn(r);
+}
+
+// CtrDrbg
+
+CAMLprim value ml_mbedtls_ctr_drbg_init() {
+	CAMLparam0();
+	CAMLlocal1(obj);
+	mbedtls_ctr_drbg_context* ctr_drbg = malloc(sizeof(mbedtls_ctr_drbg_context));
+	obj = caml_alloc(1, Abstract_tag);
+	Store_field(obj, 0, (value)ctr_drbg);
+	mbedtls_ctr_drbg_init(ctr_drbg);
+	CAMLreturn(obj);
+}
+
+CAMLprim value ml_mbedtls_ctr_drbg_random(value p_rng, value output, value output_len) {
+	CAMLparam3(p_rng, output, output_len);
+	CAMLlocal1(r);
+	r = mbedtls_ctr_drbg_random(CTR(p_rng), String_val(output), Int_val(output_len));
+	CAMLreturn(Val_int(r));
+}
+
+CAMLprim value ml_mbedtls_ctr_drbg_seed(value ctx, value p_entropy, value custom) {
+	CAMLparam2(ctx, custom);
+	CAMLlocal1(r);
+	r = mbedtls_ctr_drbg_seed(CTR(ctx), mbedtls_entropy_func, PVOID(p_entropy), NULL, 0);
+	CAMLreturn(Val_int(r));
+}
+
+// Entropy
+
+CAMLprim value ml_mbedtls_entropy_func(value data, value output, value len) {
+	CAMLparam3(data, output, len);
+	CAMLlocal1(r);
+	r = mbedtls_entropy_func(PVOID(data), String_val(output), Int_val(len));
+	CAMLreturn(Val_int(r));
+}
+
+CAMLprim value ml_mbedtls_entropy_init() {
+	CAMLparam0();
+	CAMLlocal1(obj);
+	mbedtls_entropy_context* entropy = malloc(sizeof(mbedtls_entropy_context));
+	obj = caml_alloc(1, Abstract_tag);
+	Store_field(obj, 0, (value)entropy);
+	mbedtls_entropy_init(entropy);
+	CAMLreturn(obj);
+}
+
+// Config
+
+CAMLprim void ml_mbedtls_ssl_conf_authmode(value conf, value authmode) {
+	CAMLparam2(conf, authmode);
+	mbedtls_ssl_conf_authmode(CFG(conf), Int_val(authmode));
+	CAMLreturn0;
+}
+
+CAMLprim void ml_mbedtls_ssl_conf_ca_chain(value conf, value ca_chain, value ca_crl) {
+	CAMLparam3(conf, ca_chain, ca_crl);
+	mbedtls_ssl_conf_ca_chain(CFG(conf), CERT(ca_chain), NULL /* TODO */);
```

Comment:
It's optional, so we have to handle this in the decoding.

Reply:
I have removed this for now in https://github.com/HaxeFoundation/haxe/pull/9009/commits/3174a7fa34a35737aa8c200db6422a409c5eafd7 because we can't get instances of this anyway. Will track in #9013.